File: 0003-longest-substring-without-repeating-characters/0003-longest-substring-without-repeating-characters.cpp
class Solution {
public:
    int lengthOfLongestSubstring(string s) {
        
        unordered_map<char, int> freq;
        int start=0,length=0;

        for(int end=0;end<s.size();end++)
        {
            freq[s[end]]++;
        
            while( freq[s[end]] > 1)
            {
                freq[s[start]]--;
                start++;
            }

            length = max(length,end-start+1);
        }

        return length;
    }

};

File: 0003-longest-substring-without-repeating-characters/0003-longest-substring-without-repeating-characters.py
class Solution:
    def lengthOfLongestSubstring(self, s: str) -> int:
        
        window = {}
        maxSize = 0
        l = 0
        r = 0

        while r < len(s):
            if s[r] in window:
                while s[r] in window:
                    window.pop(s[l])
                    l += 1

            window[s[r]] = r
            r += 1

            maxSize = max(maxSize,r-l)

        return maxSize
